pkg/scale/scale.go

```diff
@@ -123,16 +136,61 @@ func (cc *cruiseControlScaler) Status(ctx context.Context) (CruiseControlStatus,
 	}
 
 	return CruiseControlStatus{
-		MonitorReady:       resp.Result.MonitorState.State == types.MonitorStateRunning,
-		ExecutorReady:      resp.Result.ExecutorState.State == types.ExecutorStateTypeNoTaskInProgress,
-		AnalyzerReady:      resp.Result.AnalyzerState.IsProposalReady && goalsReady,
-		ProposalReady:      resp.Result.AnalyzerState.IsProposalReady,
+		MonitorReady:       result.MonitorState.State == types.MonitorStateRunning,
+		ExecutorReady:      result.ExecutorState.State == types.ExecutorStateTypeNoTaskInProgress,
+		AnalyzerReady:      result.AnalyzerState.IsProposalReady && goalsReady,
+		ProposalReady:      result.AnalyzerState.IsProposalReady,
 		GoalsReady:         goalsReady,
-		MonitoredWindows:   resp.Result.MonitorState.NumMonitoredWindows,
-		MonitoringCoverage: resp.Result.MonitorState.MonitoringCoveragePercentage,
+		MonitoredWindows:   result.MonitorState.NumMonitoredWindows,
+		MonitoringCoverage: result.MonitorState.MonitoringCoveragePercentage,
 	}, nil
 }
 
+func (cc *cruiseControlScaler) waitForTaskCompletion(ctx context.Context, taskID string) (*types.StateResult, error) {
+	result := &types.StateResult{}
+
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Jitter:   0,
+		Factor:   2,
+		Steps:    5,
+	}
+
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		req := &api.UserTasksRequest{
+			UserTaskIDs:         []string{taskID},
+			FetchCompletedTasks: true,
+		}
+
+		resp, err := cc.client.UserTasks(ctx, req)
+		if err != nil {
+			return false, err
+		}
+
+		if len(resp.Result.UserTasks) != 1 {
+			return false, fmt.Errorf("could not get the Cruise Control state, expected the response for 1 task (%s), but got %d responses", taskID, len(resp.Result.UserTasks))
+		}
+
+		status := resp.Result.UserTasks[0].Status
+		if status == types.UserTaskStatusCompleted {
+			if err = json.Unmarshal([]byte(resp.Result.UserTasks[0].OriginalResponse), result); err != nil {
+				return false, err
+			}
+			return true, nil
+		} else if status == types.UserTaskStatusActive || status == types.UserTaskStatusInExecution {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("could not get the Cruise Control state, the task finished with `%s` status", status)
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
```

Comment:
```suggestion
	return nil, fmt.Errorf("reached an unexpected result waiting for task (%s)", taskID)
```
Would the above make more sense? It is a little difficult to understand under what scenarios can this line be reached and what should be returned.

Reply:
Thanks for explaining that the 2nd argument to `ExponentialBackoff` is a function! Shows how well I read the code (not).